### PR TITLE
Define smart_web_messaging_token

### PR DESCRIPTION
**Pull request by jmandel**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -196,16 +205,5 @@ See [alternatives-considered.md](./alternatives-considered.md)
...
-
-2. Does access expire with SMART access_tokens?
-
-3. How does the app know what messageTypes are supported? 
-There are two aspects to this:
-* Does the container at all support messaging?
-* What messages are supported in this context?
 
-4. Does (3) above need to be an initial handshake postMessage? Do scopes in access_token already meet this need? Or do we need something like added details in a well-known/smart-configuration.json / documentation?
+* Does (3) above need to be an initial handshake postMessage? Do scopes in access_token already meet this need? Or do we need something like added details in a well-known/smart-configuration.json / documentation?
```
Comment (barabo) [inline README.md]: The `(3) above` no longer makes sense (since it has been removed).  Maybe re-word this open question to include the relevant context.

Comment (jmandel) [reply README.md]: Updated to resolve the dangling reference to (3).

**Comment by dennispatterson** [comment]
With https://github.com/smart-on-fhir/smart-web-messaging/pull/5 open simultaneously on an Implicit Authorization Context approach, it would be helpful to document some value proposition points of this approach over that one, or some pros/cons between the two.  It even says it's designed to stimulate discussion, so I'll nudge that along a bit. :)

The implicit approach is based on an EHR having knowledge of which windows correspond to which apps by virtue of opening them in a given context.  Some example questions to get started could be
* Does the exchange of a `smart_web_messaging_handle` provide additional protection against potential threats we've thought of?
* What are some examples of what the presence of a `smart_web_messaging_handle` would provide or support that implicit context alone would not?

**Comment by jmandel** [comment]
Great questions. For me:

* If an app is open in multiple windows at once and requests different scopes in each launch, there's no good way for the EHR to keep them separate. So in an implicit context, at best each instance would get access to the same set of web messaging scopes.

* There's no clear way to make sure that any one of these app instances had successfully completed a SMART launch prior to issuing Web Messaging calls. The only thing you can really be sure of is the app's origin. At best, the server would have to do some kind of time-or-ip-or-other-heuristic-based magic to try to correlate app launches with web messaging requests. I don't know how a server can really "get this right."

@dennispatterson did you manage to connect with @whitehatguy on this? Does Cerner have a preference on how to proceed?

**Comment by whitehatguy** [comment]
> If an app is open in multiple windows at once and requests different scopes in each launch, there's no good way for the EHR to keep them separate

The top level browsing context can compare the event ```source``` property of received via its message event listener to any auxiliary browsing context (window/tab) it may have opened and holds by reference, or any iframe```contentWindow``` that it may have in its child DOM.  This allows the EHR's JavaScript to determine which of those browsing contexts it created initiated the postMessage event it just received.  

> There's no clear way to make sure that any one of these app instances had successfully completed a SMART launch prior to issuing Web Messaging calls. The only thing you can really be sure of is the app's origin.

Thinking aloud - a security requirement that may be implied here is that the EHR can't technically trust the client-side code that it is launching?  Perhaps, the URL targeted for launch is some form of intermediary service that, in itself, is not trusted to act - the final destination post-authorization-code-redirect is ultimately the only trusted actor, and that its origin may be unknown at runtime to the EHR itself?  

Alternatively, are there use cases where it is expected that an application must be given the opportunity to restrict client-side invocation (example: licensing/entitlement checks)?

A simplistic use case would be one where the client-side component being invoked by the EHR is simply trusted by origin alone, scoped to whatever UI interactions should be allowed for that browsing context, and that the remote SMART on FHIR aspects (launch/authorization code flow) are kept orthogonal to that concern.  My personal ask is to understand if there are specific threats that would be unacceptable for that simplistic use case, or if there are other more valuable use cases that can only be achieved by extending the authorization code flow in the manner described here.

**Comment by jmandel** [comment]
Thanks Matt!

> The top level browsing context can compare the event source property of received via its message event listener to any auxiliary browsing context (window/tab)

Indeed, the EHR can tell the various copies of the app apart, but it doesn't have a way to associate each with a SMART launch. We may decide this is unimportant (see below) -- but if we care, then I think we need some correlation handle to tie these together.

> Thinking aloud - a security requirement that may be implied here is that the EHR can't technically trust the client-side code that it is launching?

It's a _potential_ requirement; if we leave that requirement out, things get much simpler. (The reason we might want to keep this requirement in place is to help the EHR say "we trust client-side code *if* it's being served up by an app that has completed a (possibly confidential) client launch." In other words, to build on the same set of assumptions that apply to any SMART app.

> A simplistic use case would be one where the client-side component being invoked by the EHR is simply trusted by origin alone, scoped to whatever UI interactions should be allowed for that browsing context

Scope to origin alone makes the whole game a lot simpler. In that context we'd say "we don't need a SMART launch to happen at all." I'd like help evaluating whether this is reasonable. (My sense is that if we can communicate the assumptions clearly enough, it's actually something we can run with.)

> My personal ask is to understand if there are specific threats that would be unacceptable for that simplistic use case

Thanks again for the clarity of this question. The main threat I'm aware of is something like "an attacker can serve up malicious client-side code to the EHR, and controlling the domain alone without controlling the client secret is enough to send web messaging events into the EHR." (Now, if an attacker can do this, they're very likely also able MITM the entire launch and then coax the web messaging token out of the legit server -- it's not obvious to me how an app would protect against this.)

**Comment by jmandel** [comment]
Coming back to this thread after the weekend, the things I like about an explicit token and/or handle are:

1. It re-uses existing concepts (SMART App Launch, scopes, etc)
2. It makes explicit the fact that a given web messaging request is associated with a specific app launch
3. It enables some app launches to ask for fewer scopes than others 
Re (3), the idea here is that a given app might sometimes (i.e., on some launches) need more access than others, and here it could restrict the privileges of its web messaging authorization accordingly.

This approach prevents some classes of attacks -- e.g., "xss on any page in the domain" is enough to perform a scratchpad read under the origin-only protection model; but won't necessarily work under the token-protected model. (Clearly a well-placed xss could grab a web messaging token and *then* do a scratchpad read -- but the bar is higher, since code needs to be injected into a context where the web messaging token is available, rather than just "any page on the domain".)

**Comment by whitehatguy** [comment]
> rather than just "any page on the domain"

I think the bar is only higher if developers are properly setting their content security policies to control where service workers are being installed from.  Otherwise, any XSS (that has been somehow allowed by the app's CSP policy) would be able to register a service worker and could intercept the web messaging token at any URL within that origin.

If the implication here is that a developer is properly using CSP, then there shouldn't be an XSS threat in the first place.

**Comment by jmandel** [comment]
Agreed with the assessment re: interplay of XSS + CSP, @whitehatguy. Still, most CSP policies (even when applied) aren't perfect, and aside from this class of XSS threats I still find value in (1), (2), and (3) for allowing a server to define more explicit policies that tie web messaging capabilities to app launches/scopes -- this helps limit the power of Web Messaging API calls that a given context is allowed to make. I'm struggling to articulate additional benefits, but I think there's a kind of "defense in depth" argument here.

Looking at this from the other side: are there specific reasons why we *wouldn't* want to model the web messaging token explicitly?

**Comment by jmandel** [comment]
Another thing I like about this approach is that it sets a patter for application-level authz in the event that we wind up expanding to messaging channels beyond `postMessage` (e.g., on-device native apps).

I'm going to merge this now -- in part, to see if it sparks additional discussion as we get the spec prepared for hand-off to an HL7 balloting process.

**Pull request merged**